src/backend/distributed/commands/type.c

```diff
@@ -0,0 +1,852 @@
+ *       transaction case abvove, or due to a type predating the citus
+ *       extension.
+ *     - node activation
+ *       Together with all objects that are marked as distributed in citus
+ *       types will be created during the activation of a new node to allow
+ *       reference tables to use this type.
+ *
+ * Copyright (c) 2019, Citus Data, Inc.
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/htup_details.h"
+#include "access/xact.h"
+#include "catalog/namespace.h"
+#include "catalog/pg_enum.h"
+#include "catalog/pg_type.h"
+#include "miscadmin.h"
+#include "nodes/makefuncs.h"
+#include "parser/parse_type.h"
+#include "utils/builtins.h"
+#include "utils/fmgroids.h"
+#include "utils/lsyscache.h"
+#include "utils/regproc.h"
+#include "utils/syscache.h"
+#include "utils/typcache.h"
+
+#include "distributed/dist_catalog/distobject.h"
+#include "distributed/dist_catalog/namespace.h"
+#include "distributed/commands.h"
+#include "distributed/commands/utility_hook.h"
+#include "distributed/deparser.h"
+#include "distributed/metadata_sync.h"
+#include "distributed/multi_executor.h"
+#include "distributed/relation_access_tracking.h"
+#include "distributed/remote_commands.h"
+#include "distributed/transaction_management.h"
+#include "distributed/worker_manager.h"
+#include "distributed/worker_transaction.h"
+
+#define AlterEnumIsRename(stmt) (stmt->oldVal != NULL)
+#define AlterEnumIsAddValue(stmt) (stmt->oldVal == NULL)
+
+
+#define CREATE_IF_NOT_EXISTS_COMMAND "SELECT worker_create_if_not_exists(%s);"
+#define ALTER_TYPE_OWNER_COMMAND "ALTER TYPE %s OWNER TO %s;"
+#define CREATE_OR_REPLACE_COMMAND "SELECT worker_create_or_replace(%s);"
+
+
+/* forward declaration for helper functions*/
+static void makeRangeVarQualified(RangeVar *var);
+static List * FilterNameListForDistributedTypes(List *objects);
+static List * TypeNameListToObjectAddresses(List *objects);
+static TypeName * makeTypeNameFromRangeVar(const RangeVar *relation);
+static void EnsureSequentialModeForTypeDDL(void);
+static Oid get_typowner(Oid typid);
+static const char * wrap_in_sql(const char *fmt, const char *sql);
+
+/* recreate functions */
+static CompositeTypeStmt * RecreateCompositeTypeStmt(Oid typeOid);
+static List * composite_type_coldeflist(Oid typeOid);
+static CreateEnumStmt * RecreateEnumStmt(Oid typeOid);
+static List * enum_vals_list(Oid typeOid);
+
+
+List *
+PlanCompositeTypeStmt(CompositeTypeStmt *stmt, const char *queryString)
+{
+	const char *compositeTypeStmtSql = NULL;
+	TypeName *typeName = NULL;
+	Oid typeOid = InvalidOid;
+	ObjectAddress typeAddress = { 0 };
+
+	/*
+	 * by not propagating in a transaction block we allow for parallelism to be used when
+	 * this type will be used as a column in a table that will be created and distributed
+	 * in this same transaction.
+	 */
+	if (IsTransactionBlock())
+	{
+		return NIL;
```

Comment:
How common would the following be where the user creates a type and use it in a query: not any other table

```SQL
BEGIN;
CREATE TYPE tt10 AS (a int, b int);

select (user_id, value_1)::tt10 FROM users_table LIMIT 1;
DETAIL:  on server onderkalaci@localhost:9700
ERROR:  type "public.tt10" does not exist
CONTEXT:  while executing command on localhost:9700
```

@marcocitus any comments here?

Reply:
I would guess not very common since it already errors out unless you also do run_command_on_workers, in which case it still works.